### DISCUSSION TITLE
How to create a vector with a non-copyable struct?

### DISCUSSION THREAD

**Discussion by ricardomatias** [ETLCPP/etl#822] [Q&A] [answered]
I'm trying to create the following vector of structs (with copy constructor deleted) in place and I cannot compile it. I'd like to understand why and what are the alternatives.

`etl::vector<NonCopyable, 2> { NonCopyable{1}, NonCopyable{2} }; // error`

Is it because etl::vector doesn't implement a std:initializer_list? Or because the structs are rvalues? Or some other reason?

**Comment by jwellbelove** [comment]
The ETL does support `std::initializer_list`. There are implementations for Visual Studio's compiler and GCC/Clang.

**Comment by jwellbelove** [comment]
This code doesn't work with `std::vector` either.

`Error	C2280	'NonCopyable::NonCopyable(const NonCopyable &)': attempting to reference a deleted function`

```
#include <vector>

struct NonCopyable
{
  NonCopyable(int i_)
    : i(i)
  {
  }

  NonCopyable(NonCopyable&&) = default;
  NonCopyable& operator =(NonCopyable&&) = default;

  NonCopyable(const NonCopyable&) = delete;
  NonCopyable& operator =(const NonCopyable&) = delete;

  int i;
};

int main()
{
  std::vector<NonCopyable> { NonCopyable{ 1 }, NonCopyable{ 2 } };
}
```

**Comment by jwellbelove** [comment, marked as answer]
This works.

```
#include "etl/vector.h"

struct NonCopyable
{
  NonCopyable(int i_)
    : i(i)
  {
  }

  NonCopyable(NonCopyable&&) = default;
  NonCopyable& operator =(NonCopyable&&) = default;

  NonCopyable(const NonCopyable&) = delete;
  NonCopyable& operator =(const NonCopyable&) = delete;

  int i;
};

int main()
{
  etl::vector<NonCopyable, 2> v;

  v.emplace_back(NonCopyable{ 1 });
  v.emplace_back(NonCopyable{ 2 });
}
```

**Comment by jwellbelove** [comment]
From Stackoverflow:-

`std::initializer_list` _references static memory. That's what the class is for. You cannot move from static memory, because movement implies changing it. **You can only copy from it**_

**Comment by ricardomatias** [comment]
Thank your for all your replies @jwellbelove, they're very helpful in understand this.
